src/awkward/_do.py

```diff
@@ -271,42 +271,29 @@ def mergeable(one: Content, two: Content, mergebool: bool = True) -> bool:
     return one._mergeable_next(two, mergebool=mergebool)
 
 
-def merge_as_union(one: Content, two: Content) -> ak.contents.UnionArray:
-    mylength = one.length
-    theirlength = two.length
-    tags = ak.index.Index8.empty(
-        mylength + theirlength,
-        one.backend.index_nplike,
-    )
-    index = ak.index.Index64.empty(
-        mylength + theirlength,
-        one.backend.index_nplike,
-    )
-    contents = [one, two]
-    assert tags.nplike is one.backend.index_nplike
-    one._handle_error(
-        one.backend["awkward_UnionArray_filltags_const", tags.dtype.type](
-            tags.data, 0, mylength, 0
-        )
-    )
-    assert index.nplike is one.backend.index_nplike
-    one._handle_error(
-        one.backend["awkward_UnionArray_fillindex_count", index.dtype.type](
-            index.data, 0, mylength
-        )
-    )
-    one._handle_error(
-        one.backend["awkward_UnionArray_filltags_const", tags.dtype.type](
-            tags.data, mylength, theirlength, 1
+def merge_as_union(
+    contents: Sequence[Content], parameters=None
+) -> ak.contents.UnionArray:
+    length = sum([c.length for c in contents])
+    first = contents[0]
+    tags = ak.index.Index8.empty(length, first.backend.index_nplike)
+    index = ak.index.Index64.empty(length, first.backend.index_nplike)
+
+    offset = 0
+    for i, content in enumerate(contents):
+        content._handle_error(
+            content.backend["awkward_UnionArray_filltags_const", tags.dtype.type](
+                tags.data, offset, content.length, i
+            )
         )
-    )
-    one._handle_error(
-        one.backend["awkward_UnionArray_fillindex_count", index.dtype.type](
-            index.data, mylength, theirlength
+        content._handle_error(
+            content.backend["awkward_UnionArray_fillindex_count", index.dtype.type](
+                index.data, offset, content.length
+            )
         )
-    )
+        offset += content.length
 
-    return ak.contents.UnionArray(tags, index, contents, parameters=None)
+    return ak.contents.UnionArray(tags, index, contents, parameters=parameters)
```

Comment:
@jpivarski this PR currently fails part of the test suite. The reason is that our many-batch merging scheme can produce batches whose non-adjacent merge results ("partial merges") can themselves be further merged. We discussed during our meeting that this is a possibility, yet our simplification tests check:
```python
for content in contents[1:]:
    if content._mergeable_next(contents[0], ...):
        raise RuntimeError
```
and our simplification algorithm performs an exhaustive `N^2` loop

Interestingly, this means it is actually possible to produce non-canonical union arrays, e.g. if `B` and `C` are mergeable in `[A, B, C]`, we wouldn't detect it in the initialiser. I don't have a solution here, I just wanted to raise it.

For this particular PR, the easiest fix would again be to use `UnionArray.simplified`, which will ensure that any mergeable partial merges are indeed merged. We could also implement our own `N(N-1)/2` algorithm that performs batch merging, but it's diminishing returns. A nudge in the right direction would be nice.

Reply:
We discussed that this should be done, but outside of the initialiser.